// @sora-substrate/util/build/staking/helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.formatValidatorExposure = exports.formatPayee = exports.formatNominations = exports.formatIndividualRewardPoints = exports.formatEra = void 0;
var _types = require("./types");
const formatEra = data => {
  const era = data.unwrap();
  return era.toNumber();
};
exports.formatEra = formatEra;
const formatIndividualRewardPoints = data => {
  const result = {};
  for (const [account, points] of data.individual.entries()) {
    result[account.toString()] = points.toNumber();
  }
  return result;
};
exports.formatIndividualRewardPoints = formatIndividualRewardPoints;
const formatNominations = codec => {
  if (codec.isEmpty) return null;
  const data = codec.unwrap();
  const targets = data.targets.map(target => target.toString());
  const suppressed = data.suppressed.isTrue;
  const submittedIn = data.submittedIn.toNumber();
  return {
    targets,
    suppressed,
    submittedIn
  };
};
exports.formatNominations = formatNominations;
const formatPayee = payee => {
  return payee in _types.StakingRewardsDestination ? payee : {
    Account: payee
  };
};
exports.formatPayee = formatPayee;
const formatValidatorExposure = codec => {
  return {
    total: codec.total.toString(),
    own: codec.own.toString(),
    others: codec.others.map(item => ({
      who: item.who.toString(),
      value: item.value.toString()
    }))
  };
};
exports.formatValidatorExposure = formatValidatorExposure;
