// @sora-substrate/util/build/bridgeProxy/evm/consts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EvmNetworkId = exports.EvmAppKinds = void 0;
let EvmNetworkId;
exports.EvmNetworkId = EvmNetworkId;
(function (EvmNetworkId) {
  EvmNetworkId[EvmNetworkId["EthereumMainnet"] = 1] = "EthereumMainnet";
  EvmNetworkId[EvmNetworkId["EthereumRopsten"] = 3] = "EthereumRopsten";
  EvmNetworkId[EvmNetworkId["EthereumRinkeby"] = 4] = "EthereumRinkeby";
  EvmNetworkId[EvmNetworkId["EthereumGoerli"] = 5] = "EthereumGoerli";
  EvmNetworkId[EvmNetworkId["EthereumKovan"] = 42] = "EthereumKovan";
  EvmNetworkId[EvmNetworkId["EthereumSepolia"] = 11155111] = "EthereumSepolia";
  EvmNetworkId[EvmNetworkId["BinanceSmartChainMainnet"] = 56] = "BinanceSmartChainMainnet";
  EvmNetworkId[EvmNetworkId["BinanceSmartChainTestnet"] = 97] = "BinanceSmartChainTestnet";
  EvmNetworkId[EvmNetworkId["EthereumClassicMainnet"] = 61] = "EthereumClassicMainnet";
  EvmNetworkId[EvmNetworkId["EthereumClassicTestnetMordor"] = 63] = "EthereumClassicTestnetMordor";
  EvmNetworkId[EvmNetworkId["PolygonMainnet"] = 137] = "PolygonMainnet";
  EvmNetworkId[EvmNetworkId["PolygonTestnetMumbai"] = 80001] = "PolygonTestnetMumbai";
  EvmNetworkId[EvmNetworkId["KlaytnMainnet"] = 8217] = "KlaytnMainnet";
  EvmNetworkId[EvmNetworkId["KlaytnTestnetBaobab"] = 1001] = "KlaytnTestnetBaobab";
  EvmNetworkId[EvmNetworkId["AvalancheMainnet"] = 43114] = "AvalancheMainnet";
  EvmNetworkId[EvmNetworkId["AvalancheTestnetFuji"] = 43113] = "AvalancheTestnetFuji";
  EvmNetworkId[EvmNetworkId["ArbitrumMainnet"] = 42161] = "ArbitrumMainnet";
  EvmNetworkId[EvmNetworkId["ArbitrumGoerliTestnet"] = 421613] = "ArbitrumGoerliTestnet";
  EvmNetworkId[EvmNetworkId["FantomMainnet"] = 250] = "FantomMainnet";
  EvmNetworkId[EvmNetworkId["FantomTestnet"] = 4002] = "FantomTestnet";
})(EvmNetworkId || (exports.EvmNetworkId = EvmNetworkId = {}));
let EvmAppKinds;
exports.EvmAppKinds = EvmAppKinds;
(function (EvmAppKinds) {
  EvmAppKinds["EthApp"] = "EthApp";
  EvmAppKinds["ERC20App"] = "ERC20App";
  EvmAppKinds["SidechainApp"] = "SidechainApp";
  EvmAppKinds["HashiBridge"] = "HashiBridge";
  EvmAppKinds["XorMaster"] = "XorMaster";
  EvmAppKinds["ValMaster"] = "ValMaster";
})(EvmAppKinds || (exports.EvmAppKinds = EvmAppKinds = {}));
