// @sora-substrate/types/build/interfaces/definitions.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "assets", {
  enumerable: true,
  get: function () {
    return _definitions.default;
  }
});
Object.defineProperty(exports, "basicChannel", {
  enumerable: true,
  get: function () {
    return _definitions21.default;
  }
});
Object.defineProperty(exports, "bridgeProxy", {
  enumerable: true,
  get: function () {
    return _definitions24.default;
  }
});
Object.defineProperty(exports, "ceresGovernancePlatform", {
  enumerable: true,
  get: function () {
    return _definitions5.default;
  }
});
Object.defineProperty(exports, "ceresLaunchpad", {
  enumerable: true,
  get: function () {
    return _definitions6.default;
  }
});
Object.defineProperty(exports, "ceresLiquidityLocker", {
  enumerable: true,
  get: function () {
    return _definitions3.default;
  }
});
Object.defineProperty(exports, "ceresStaking", {
  enumerable: true,
  get: function () {
    return _definitions2.default;
  }
});
Object.defineProperty(exports, "ceresTokenLocker", {
  enumerable: true,
  get: function () {
    return _definitions4.default;
  }
});
Object.defineProperty(exports, "demeterFarmingPlatform", {
  enumerable: true,
  get: function () {
    return _definitions18.default;
  }
});
Object.defineProperty(exports, "dexApi", {
  enumerable: true,
  get: function () {
    return _definitions7.default;
  }
});
Object.defineProperty(exports, "dexManager", {
  enumerable: true,
  get: function () {
    return _definitions8.default;
  }
});
Object.defineProperty(exports, "ethBridge", {
  enumerable: true,
  get: function () {
    return _definitions9.default;
  }
});
Object.defineProperty(exports, "farming", {
  enumerable: true,
  get: function () {
    return _definitions10.default;
  }
});
Object.defineProperty(exports, "hermesGovernancePlatform", {
  enumerable: true,
  get: function () {
    return _definitions23.default;
  }
});
Object.defineProperty(exports, "intentivizedChannel", {
  enumerable: true,
  get: function () {
    return _definitions22.default;
  }
});
Object.defineProperty(exports, "irohaMigration", {
  enumerable: true,
  get: function () {
    return _definitions11.default;
  }
});
Object.defineProperty(exports, "leafProvider", {
  enumerable: true,
  get: function () {
    return _definitions20.default;
  }
});
Object.defineProperty(exports, "liquidityProxy", {
  enumerable: true,
  get: function () {
    return _definitions12.default;
  }
});
Object.defineProperty(exports, "pswapDistribution", {
  enumerable: true,
  get: function () {
    return _definitions13.default;
  }
});
Object.defineProperty(exports, "rewards", {
  enumerable: true,
  get: function () {
    return _definitions14.default;
  }
});
Object.defineProperty(exports, "runtime", {
  enumerable: true,
  get: function () {
    return _definitions15.default;
  }
});
Object.defineProperty(exports, "template", {
  enumerable: true,
  get: function () {
    return _definitions16.default;
  }
});
Object.defineProperty(exports, "tradingPair", {
  enumerable: true,
  get: function () {
    return _definitions17.default;
  }
});
Object.defineProperty(exports, "vestedRewards", {
  enumerable: true,
  get: function () {
    return _definitions19.default;
  }
});
var _definitions = _interopRequireDefault(require("./assets/definitions"));
var _definitions2 = _interopRequireDefault(require("./ceresStaking/definitions"));
var _definitions3 = _interopRequireDefault(require("./ceresLiquidityLocker/definitions"));
var _definitions4 = _interopRequireDefault(require("./ceresTokenLocker/definitions"));
var _definitions5 = _interopRequireDefault(require("./ceresGovernancePlatform/definitions"));
var _definitions6 = _interopRequireDefault(require("./ceresLaunchpad/definitions"));
var _definitions7 = _interopRequireDefault(require("./dexApi/definitions"));
var _definitions8 = _interopRequireDefault(require("./dexManager/definitions"));
var _definitions9 = _interopRequireDefault(require("./ethBridge/definitions"));
var _definitions10 = _interopRequireDefault(require("./farming/definitions"));
var _definitions11 = _interopRequireDefault(require("./irohaMigration/definitions"));
var _definitions12 = _interopRequireDefault(require("./liquidityProxy/definitions"));
var _definitions13 = _interopRequireDefault(require("./pswapDistribution/definitions"));
var _definitions14 = _interopRequireDefault(require("./rewards/definitions"));
var _definitions15 = _interopRequireDefault(require("./runtime/definitions"));
var _definitions16 = _interopRequireDefault(require("./template/definitions"));
var _definitions17 = _interopRequireDefault(require("./tradingPair/definitions"));
var _definitions18 = _interopRequireDefault(require("./demeterFarmingPlatform/definitions"));
var _definitions19 = _interopRequireDefault(require("./vestedRewards/definitions"));
var _definitions20 = _interopRequireDefault(require("./leafProvider/definitions"));
var _definitions21 = _interopRequireDefault(require("./basicChannel/definitions"));
var _definitions22 = _interopRequireDefault(require("./intentivizedChannel/definitions"));
var _definitions23 = _interopRequireDefault(require("./hermesGovernancePlatform/definitions"));
var _definitions24 = _interopRequireDefault(require("./bridgeProxy/definitions"));
