// @sora-substrate/util/build/bridgeProxy/consts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BridgeTxStatus = exports.BridgeTxDirection = exports.BridgeNetworkType = exports.BridgeAccountType = void 0;
let BridgeTxDirection;
exports.BridgeTxDirection = BridgeTxDirection;
(function (BridgeTxDirection) {
  BridgeTxDirection["Outgoing"] = "Outgoing";
  BridgeTxDirection["Incoming"] = "Incoming";
})(BridgeTxDirection || (exports.BridgeTxDirection = BridgeTxDirection = {}));
let BridgeTxStatus;
exports.BridgeTxStatus = BridgeTxStatus;
(function (BridgeTxStatus) {
  BridgeTxStatus["Done"] = "Done";
  BridgeTxStatus["Failed"] = "Failed";
  BridgeTxStatus["Pending"] = "Pending";
  BridgeTxStatus["Ready"] = "ApprovalsReady";
  BridgeTxStatus["Frozen"] = "Frozen";
  BridgeTxStatus["Broken"] = "Broken";
})(BridgeTxStatus || (exports.BridgeTxStatus = BridgeTxStatus = {}));
let BridgeNetworkType;
exports.BridgeNetworkType = BridgeNetworkType;
(function (BridgeNetworkType) {
  BridgeNetworkType["Eth"] = "EVMLegacy";
  BridgeNetworkType["Evm"] = "EVM";
  BridgeNetworkType["Sub"] = "Sub";
})(BridgeNetworkType || (exports.BridgeNetworkType = BridgeNetworkType = {}));
let BridgeAccountType;
exports.BridgeAccountType = BridgeAccountType;
(function (BridgeAccountType) {
  BridgeAccountType["Evm"] = "EVM";
  BridgeAccountType["Sora"] = "Sora";
  BridgeAccountType["Parachain"] = "Parachain";
  BridgeAccountType["Unknown"] = "Unknown";
})(BridgeAccountType || (exports.BridgeAccountType = BridgeAccountType = {}));
