// @sora-substrate/util/build/bridgeProxy/sub/consts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.XcmVersionedMultiLocation = exports.XcmV3NetworkId = exports.XcmV3BodyId = exports.XcmV2NetworkId = exports.XcmV2BodyId = exports.XcmMultilocationJunction = exports.XcmJunction = exports.XcmBodyPart = exports.SubNetwork = exports.SubAssetKind = void 0;
let SubAssetKind;
exports.SubAssetKind = SubAssetKind;
(function (SubAssetKind) {
  SubAssetKind["Thischain"] = "Thischain";
  SubAssetKind["Sidechain"] = "Sidechain";
})(SubAssetKind || (exports.SubAssetKind = SubAssetKind = {}));
let SubNetwork;
exports.SubNetwork = SubNetwork;
(function (SubNetwork) {
  SubNetwork["Kusama"] = "Kusama";
  SubNetwork["Polkadot"] = "Polkadot";
  SubNetwork["Rococo"] = "Rococo";
  SubNetwork["Custom"] = "Custom";
  SubNetwork["RococoSora"] = "RococoSora";
  SubNetwork["KusamaKarura"] = "KusamaKarura";
  SubNetwork["KusamaSora"] = "KusamaSora";
})(SubNetwork || (exports.SubNetwork = SubNetwork = {}));
let XcmVersionedMultiLocation; // V2 & V3
exports.XcmVersionedMultiLocation = XcmVersionedMultiLocation;
(function (XcmVersionedMultiLocation) {
  XcmVersionedMultiLocation["V2"] = "V2";
  XcmVersionedMultiLocation["V3"] = "V3";
})(XcmVersionedMultiLocation || (exports.XcmVersionedMultiLocation = XcmVersionedMultiLocation = {}));
let XcmMultilocationJunction; // V2 & V3
exports.XcmMultilocationJunction = XcmMultilocationJunction;
(function (XcmMultilocationJunction) {
  XcmMultilocationJunction["X1"] = "X1";
  XcmMultilocationJunction["X2"] = "X2";
  XcmMultilocationJunction["X3"] = "X3";
  XcmMultilocationJunction["X4"] = "X4";
  XcmMultilocationJunction["X5"] = "X5";
  XcmMultilocationJunction["X6"] = "X6";
  XcmMultilocationJunction["X7"] = "X7";
  XcmMultilocationJunction["X8"] = "X8";
  XcmMultilocationJunction["Here"] = "Here";
})(XcmMultilocationJunction || (exports.XcmMultilocationJunction = XcmMultilocationJunction = {}));
let XcmJunction; // XcmV2NetworkId
exports.XcmJunction = XcmJunction;
(function (XcmJunction) {
  XcmJunction["Parachain"] = "Parachain";
  XcmJunction["AccountId32"] = "AccountId32";
  XcmJunction["AccountIndex64"] = "AccountIndex64";
  XcmJunction["AccountKey20"] = "AccountKey20";
  XcmJunction["PalletInstance"] = "PalletInstance";
  XcmJunction["GeneralIndex"] = "GeneralIndex";
  XcmJunction["GeneralKey"] = "GeneralKey";
  XcmJunction["OnlyChild"] = "OnlyChild";
  XcmJunction["Plurality"] = "Plurality";
  XcmJunction["GlobalConsensus"] = "GlobalConsensus";
})(XcmJunction || (exports.XcmJunction = XcmJunction = {}));
let XcmV2NetworkId; // XcmV3JunctionNetworkId
exports.XcmV2NetworkId = XcmV2NetworkId;
(function (XcmV2NetworkId) {
  XcmV2NetworkId["Any"] = "Any";
  XcmV2NetworkId["Named"] = "Named";
  XcmV2NetworkId["Polkadot"] = "Polkadot";
  XcmV2NetworkId["Kusama"] = "Kusama";
})(XcmV2NetworkId || (exports.XcmV2NetworkId = XcmV2NetworkId = {}));
let XcmV3NetworkId; // XcmV2BodyId
exports.XcmV3NetworkId = XcmV3NetworkId;
(function (XcmV3NetworkId) {
  XcmV3NetworkId["ByGenesis"] = "ByGenesis";
  XcmV3NetworkId["ByFork"] = "ByFork";
  XcmV3NetworkId["Polkadot"] = "Polkadot";
  XcmV3NetworkId["Kusama"] = "Kusama";
  XcmV3NetworkId["Westend"] = "Westend";
  XcmV3NetworkId["Rococo"] = "Rococo";
  XcmV3NetworkId["Wococo"] = "Wococo";
  XcmV3NetworkId["Ethereum"] = "Ethereum";
  XcmV3NetworkId["BitcoinCore"] = "BitcoinCore";
  XcmV3NetworkId["BitcoinCash"] = "BitcoinCash";
})(XcmV3NetworkId || (exports.XcmV3NetworkId = XcmV3NetworkId = {}));
let XcmV2BodyId; // XcmV3JunctionBodyId
exports.XcmV2BodyId = XcmV2BodyId;
(function (XcmV2BodyId) {
  XcmV2BodyId["Unit"] = "Unit";
  XcmV2BodyId["Named"] = "Named";
  XcmV2BodyId["Index"] = "Index";
  XcmV2BodyId["Executive"] = "Executive";
  XcmV2BodyId["Technical"] = "Technical";
  XcmV2BodyId["Legislative"] = "Legislative";
  XcmV2BodyId["Judicial"] = "Judicial";
  XcmV2BodyId["Defense"] = "Defense";
  XcmV2BodyId["Administration"] = "Administration";
  XcmV2BodyId["Treasury"] = "Treasury";
})(XcmV2BodyId || (exports.XcmV2BodyId = XcmV2BodyId = {}));
let XcmV3BodyId; // XcmV2BodyPart
// XcmV3JunctionBodyPart
exports.XcmV3BodyId = XcmV3BodyId;
(function (XcmV3BodyId) {
  XcmV3BodyId["Unit"] = "Unit";
  XcmV3BodyId["Moniker"] = "Moniker";
  XcmV3BodyId["Index"] = "Index";
  XcmV3BodyId["Executive"] = "Executive";
  XcmV3BodyId["Technical"] = "Technical";
  XcmV3BodyId["Legislative"] = "Legislative";
  XcmV3BodyId["Judicial"] = "Judicial";
  XcmV3BodyId["Defense"] = "Defense";
  XcmV3BodyId["Administration"] = "Administration";
  XcmV3BodyId["Treasury"] = "Treasury";
})(XcmV3BodyId || (exports.XcmV3BodyId = XcmV3BodyId = {}));
let XcmBodyPart;
exports.XcmBodyPart = XcmBodyPart;
(function (XcmBodyPart) {
  XcmBodyPart["Voice"] = "Voice";
  XcmBodyPart["Members"] = "Members";
  XcmBodyPart["Fraction"] = "Fraction";
  XcmBodyPart["AtLeastProportion"] = "AtLeastProportion";
  XcmBodyPart["MoreThanProportion"] = "MoreThanProportion";
})(XcmBodyPart || (exports.XcmBodyPart = XcmBodyPart = {}));
