// @sora-substrate/liquidity-proxy/build/quote/tbc.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tbcSellPriceNoVolume = exports.tbcQuoteWithoutImpact = exports.tbcQuote = exports.tbcBuyPriceNoVolume = void 0;
var _math = require("@sora-substrate/math");
var _consts = require("../consts");
var _utils = require("../utils");
var _price = require("./price");
/**
 * This function is used to determine particular asset price in terms of a reference asset, which is set for
 * bonding curve (there could be only single token chosen as reference for all comparisons). Basically, the
 * reference token is expected to be a USD-bound stablecoin, e.g. DAI.
 *
 * Example use: understand actual value of two tokens in terms of USD.
 */
const tbcReferencePrice = (assetAddress, priceVariant, payload) => {
  const referenceAssetId = payload.consts.tbc.referenceAsset;
  // always treat TBCD as being worth $1
  if ([referenceAssetId, _consts.Consts.TBCD].includes(assetAddress)) {
    return _math.FPNumber.ONE;
  } else {
    return (0, _price.getAveragePrice)(assetAddress, referenceAssetId, priceVariant, payload);
  }
};

/**
 * Calculate USD price for single collateral asset that is stored in reserves account. In other words, find out how much
 * reserves worth, considering only one asset type.
 */
const actualReservesReferencePrice = (collateralAsset, payload, priceVariant) => {
  const reserve = _math.FPNumber.fromCodecValue(payload.reserves.tbc[collateralAsset]);
  const price = tbcReferencePrice(collateralAsset, priceVariant, payload);
  return reserve.mul(price);
};

/**
 * Calculate USD price for all XOR in network, this is done by applying ideal sell function to XOR total supply.
 * `delta` is a XOR supply offset from current total supply.
 *
 * `((initial_price + current_state) / 2) * (xor_issuance + delta)`
 */
const idealReservesReferencePrice = (collateralAssetId, priceVariant, delta, payload) => {
  const xorIssuance = _math.FPNumber.fromCodecValue(payload.issuances[_consts.Consts.XOR]);
  const initialPrice = _math.FPNumber.fromCodecValue(payload.consts.tbc.initialPrice);
  const currentState = tbcBuyFunction(collateralAssetId, priceVariant, delta, payload);
  return (0, _utils.safeDivide)(initialPrice.add(currentState), new _math.FPNumber(2)).mul(xorIssuance.add(delta));
};

/**
 * Mapping that defines ratio of fee penalty applied for selling XOR with
 * low collateralized reserves.
 */
const mapCollateralizedFractionToPenalty = fraction => {
  if (_math.FPNumber.isLessThan(fraction, new _math.FPNumber(0.05))) {
    return new _math.FPNumber(0.09);
  } else if (_math.FPNumber.isGreaterThanOrEqualTo(fraction, new _math.FPNumber(0.05)) && _math.FPNumber.isLessThan(fraction, new _math.FPNumber(0.1))) {
    return new _math.FPNumber(0.06);
  } else if (_math.FPNumber.isGreaterThanOrEqualTo(fraction, new _math.FPNumber(0.1)) && _math.FPNumber.isLessThan(fraction, new _math.FPNumber(0.2))) {
    return new _math.FPNumber(0.03);
  } else if (_math.FPNumber.isGreaterThanOrEqualTo(fraction, new _math.FPNumber(0.2)) && _math.FPNumber.isLessThan(fraction, new _math.FPNumber(0.3))) {
    return new _math.FPNumber(0.01);
  } else {
    return _math.FPNumber.ZERO;
  }
};

/**
 * Calculate ratio of fee penalty that is applied to trades when XOR is sold while
 * reserves are low for target collateral asset.
 */
const sellPenalty = (collateralAsset, payload) => {
  const idealReservesPrice = idealReservesReferencePrice(collateralAsset, _consts.PriceVariant.Sell, _math.FPNumber.ZERO, payload);
  const collateralReservesPrice = actualReservesReferencePrice(collateralAsset, payload, _consts.PriceVariant.Sell);
  if (collateralReservesPrice.isZero()) {
    throw new Error(`[liquidityProxy] TBC: Not enough collateral reserves ${collateralAsset}`);
  }
  const collateralizedFraction = (0, _utils.safeDivide)(collateralReservesPrice, idealReservesPrice);
  const penalty = mapCollateralizedFractionToPenalty(collateralizedFraction);
  return penalty;
};

/**
 * Calculate amount of PSWAP rewarded for collateralizing XOR in TBC.
 *
 * `ideal_reserves_before = sell_function(0 to xor_total_supply_before_trade)`
 *
 * `ideal_reserves_after = sell_function(0 to xor_total_supply_after_trade)`
 *
 * `actual_reserves_before = collateral_asset_reserves * collateral_asset_usd_price`
 *
 * `actual_reserves_after = actual_reserves_before + collateral_asset_input_amount * collateral_asset_usd_price`
 *
 * `unfunded_liabilities = (ideal_reserves_before - actual_reserves_before)`
 *
 * `a = unfunded_liabilities / ideal_reserves_before`
 *
 * `b = unfunded_liabilities / ideal_reserves_after`
 *
 * `P = initial_pswap_rewards`
 *
 * `N = enabled reserve currencies except PSWAP and VAL`
 *
 * `reward_pswap = ((a - b) * mean(a, b) * P) / N`
 */
const checkRewards = (collateralAsset, xorAmount, payload) => {
  if ([_consts.Consts.PSWAP, _consts.Consts.VAL, _consts.Consts.XST, _consts.Consts.TBCD].includes(collateralAsset)) {
    return [];
  }
  const idealBefore = idealReservesReferencePrice(collateralAsset, _consts.PriceVariant.Buy, _math.FPNumber.ZERO, payload);
  const idealAfter = idealReservesReferencePrice(collateralAsset, _consts.PriceVariant.Buy, xorAmount, payload);
  const actualBefore = actualReservesReferencePrice(collateralAsset, payload, _consts.PriceVariant.Buy);
  const unfundedLiabilities = idealBefore.sub(actualBefore);
  const a = (0, _utils.safeDivide)(unfundedLiabilities, idealBefore);
  const b = (0, _utils.safeDivide)(unfundedLiabilities, idealAfter);
  const mean = (0, _utils.safeDivide)(a.add(b), new _math.FPNumber(2));
  const amount = (0, _utils.safeDivide)(a.sub(b).mul(_consts.Consts.initialPswapTbcRewardsAmount).mul(mean), _consts.Consts.incentivizedCurrenciesNum);
  if (amount.isZero()) {
    return [];
  }
  return [{
    amount: amount.toCodecString(),
    currency: _consts.Consts.PSWAP,
    reason: _consts.RewardReason.BuyOnBondingCurve
  }];
};

/**
 * Buy function with regards to asset total supply and its change delta. It represents the amount of
 * input collateral required from the User to receive the requested XOR amount, i.e., the price the User buys XOR at
 * XOR is also referred as main asset.
 * Value of `delta` is assumed to be either positive or negative.
 * For every `price_change_step` tokens the price goes up by `price_change_rate`.
 *
 * `buy_price_usd = (xor_total_supply + xor_supply_delta) / (price_change_step * price_change_rate) + initial_price_usd`
 */
const tbcBuyFunction = (collateralAssetId, priceVariant, delta, payload) => {
  if (collateralAssetId === _consts.Consts.TBCD) {
    // Handle TBCD
    const xp = tbcReferencePrice(_consts.Consts.XOR, priceVariant, payload);
    // get the XOR price in USD (DAI) and add $1 to it
    const xorPrice = xp.add(_math.FPNumber.ONE);
    return xorPrice;
  } else {
    // Everything other than TBCD
    const xorIssuance = _math.FPNumber.fromCodecValue(payload.issuances[_consts.Consts.XOR]);
    const initialPrice = _math.FPNumber.fromCodecValue(payload.consts.tbc.initialPrice);
    const priceChangeStep = _math.FPNumber.fromCodecValue(payload.consts.tbc.priceChangeStep);
    const priceChangeRate = _math.FPNumber.fromCodecValue(payload.consts.tbc.priceChangeRate);
    return (0, _utils.safeDivide)(xorIssuance.add(delta), priceChangeStep.mul(priceChangeRate)).add(initialPrice);
  }
};

/**
 * Sell function with regards to asset total supply and its change delta. It represents the amount of
 * output collateral tokens received by User by indicating exact sold XOR amount. I.e. the price User sells at.
 * Value of `delta` is assumed to be either positive or negative.
 * Sell function is `sell_price_coefficient`% of buy function (see `tbcBuyFunction`).
 *
 * `sell_price = sell_price_coefficient * buy_price`
 */
const tbcSellFunction = (collateralAssetId, delta, payload) => {
  const buyFunctionResult = tbcBuyFunction(collateralAssetId, _consts.PriceVariant.Sell, delta, payload);
  const sellPriceCoefficient = _math.FPNumber.fromCodecValue(payload.consts.tbc.sellPriceCoefficient);
  return buyFunctionResult.mul(sellPriceCoefficient);
};

/**
 * Calculates and returns the current sell price, assuming that input is the main asset and output is the collateral asset.
 * To calculate sell price for a specific amount of assets:
 * 1. Current reserves of collateral token are taken
 * 2. Same amount by value is assumed for main asset
 *
 *    2.1. Values are compared via getting prices for both main and collateral tokens with regard to another token called reference token which is set for particular pair. This should be e.g. stablecoin DAI.
 *
 *    2.2. Reference price for base token is taken as 80% of current bonding curve buy price.
 *
 *    2.3. Reference price for collateral token is taken as current market price, i.e. price for 1 token on liquidity proxy.
 *
 * 3. Given known reserves for main and collateral, output collateral amount is calculated by applying x*y=k model resulting in curve-like dependency.
 */
const tbcSellPrice = (collateralAsset, amount, isDesiredInput, payload) => {
  const collateralSupply = _math.FPNumber.fromCodecValue(payload.reserves.tbc[collateralAsset]);
  const mainPricePerReferenceUnit = tbcSellFunction(collateralAsset, _math.FPNumber.ZERO, payload);
  const collateralPricePerReferenceUnit = tbcReferencePrice(collateralAsset, _consts.PriceVariant.Sell, payload);
  const mainSupply = (0, _utils.safeDivide)(collateralSupply.mul(collateralPricePerReferenceUnit), mainPricePerReferenceUnit);
  if (isDesiredInput) {
    const outputCollateral = (0, _utils.safeDivide)(amount.mul(collateralSupply), mainSupply.add(amount));
    if (_math.FPNumber.isGreaterThan(outputCollateral, collateralSupply)) {
      throw new Error(`[liquidityProxy] TBC: Not enough collateral reserves ${collateralAsset}`);
    }
    return outputCollateral;
  } else {
    if (_math.FPNumber.isGreaterThan(amount, collateralSupply)) {
      throw new Error(`[liquidityProxy] TBC: Not enough collateral reserves ${collateralAsset}`);
    }
    const outputXor = (0, _utils.safeDivide)(mainSupply.mul(amount), collateralSupply.sub(amount));
    return outputXor;
  }
};

/**
Calculates and returns the current buy price, assuming that input is the collateral asset and output is the main asset.

To calculate price for a specific amount of assets (with desired main asset output),
one needs to calculate the area of a right trapezoid.

`AB` : buy_function(xor_total_supply)
`CD` : buy_function(xor_total_supply + xor_supply_delta)

```nocompile
         ..  C
       ..  │
  B  ..    │
    │   S  │
    │      │
  A └──────┘ D
```

1) Amount of collateral tokens needed in USD to get `xor_supply_delta`(AD) XOR tokens
S = ((AB + CD) / 2) * AD

or

buy_price_usd = ((buy_function(xor_total_supply) + buy_function(xor_total_supply + xor_supply_delta)) / 2) * xor_supply_delta

2) Amount of XOR tokens received by depositing `S` collateral tokens in USD:

Solving right trapezoid area formula with respect to `xor_supply_delta` (AD),
actual square `S` is known and represents collateral amount.
We have a quadratic equation:

buy_function(x) = price_change_coefficient * x + initial_price

Assume `P` = price_change_coefficient = 1337

M * AD² + 2 * AB * AD - 2 * S = 0
equation with two solutions, taking only positive one:
AD = (√((AB * 2 / M)² + 8 * S / M) - 2 * AB / M) / 2 (old formula)
AD = √(P * (AB² * P + 2 * S)) - AB * P (new formula)

or
(old)
xor_supply_delta = (√((buy_function(xor_total_supply) * 2 / price_change_coeff)²
  + 8 * buy_price_usd / price_change_coeff) - 2 * buy_function(xor_total_supply)
 / price_change_coeff) / 2
(new)
xor_supply_delta = √price_change_coefficient * √(buy_function(xor_total_supply)² * price_change_coefficient + 2 * buy_price_usd)
  - buy_function(xor_total_supply) * price_change_coefficient
 */
const tbcBuyPrice = (collateralAsset, amount, isDesiredInput, payload) => {
  const priceChangeStep = _math.FPNumber.fromCodecValue(payload.consts.tbc.priceChangeStep);
  const priceChangeRate = _math.FPNumber.fromCodecValue(payload.consts.tbc.priceChangeRate);
  const priceChangeCoeff = priceChangeStep.mul(priceChangeRate);
  const currentState = tbcBuyFunction(collateralAsset, _consts.PriceVariant.Buy, _math.FPNumber.ZERO, payload);
  const collateralPricePerReferenceUnit = tbcReferencePrice(collateralAsset, _consts.PriceVariant.Buy, payload);
  if (isDesiredInput) {
    const collateralReferenceIn = collateralPricePerReferenceUnit.mul(amount);
    let mainOut;
    if (collateralAsset === _consts.Consts.TBCD) {
      mainOut = (0, _utils.safeDivide)(collateralReferenceIn, currentState);
    } else {
      const sqrt = currentState.mul(currentState).mul(priceChangeCoeff).add(new _math.FPNumber(2).mul(collateralReferenceIn)).mul(priceChangeCoeff).sqrt();
      mainOut = sqrt.sub(currentState.mul(priceChangeCoeff));
    }
    return (0, _utils.getMaxPositive)(mainOut);
  } else {
    const newState = tbcBuyFunction(collateralAsset, _consts.PriceVariant.Buy, amount, payload);
    const collateralReferenceIn = (0, _utils.safeDivide)(currentState.add(newState).mul(amount), new _math.FPNumber(2));
    const collateralQuantity = (0, _utils.safeDivide)(collateralReferenceIn, collateralPricePerReferenceUnit);
    return (0, _utils.getMaxPositive)(collateralQuantity);
  }
};

// decide_sell_amounts
const tbcSellPriceWithFee = function (collateralAsset, amount, isDesiredInput, payload) {
  let deduceFee = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : true;
  const newFee = deduceFee ? _consts.Consts.TBC_FEE.add(sellPenalty(collateralAsset, payload)) : _math.FPNumber.ZERO;
  if (isDesiredInput) {
    const fee = amount.mul(newFee);
    const outputAmount = tbcSellPrice(collateralAsset, amount.sub(fee), isDesiredInput, payload);
    return {
      amount: outputAmount,
      fee,
      rewards: [],
      distribution: [{
        input: _consts.Consts.XOR,
        output: collateralAsset,
        market: _consts.LiquiditySourceTypes.MulticollateralBondingCurvePool,
        income: amount,
        outcome: outputAmount,
        fee
      }]
    };
  } else {
    const inputAmount = tbcSellPrice(collateralAsset, amount, isDesiredInput, payload);
    const inputAmountWithFee = (0, _utils.safeDivide)(inputAmount, _math.FPNumber.ONE.sub(newFee));
    const fee = inputAmountWithFee.sub(inputAmount);
    return {
      amount: inputAmountWithFee,
      fee,
      rewards: [],
      distribution: [{
        input: _consts.Consts.XOR,
        output: collateralAsset,
        market: _consts.LiquiditySourceTypes.MulticollateralBondingCurvePool,
        income: inputAmountWithFee,
        outcome: amount,
        fee
      }]
    };
  }
};
const tbcBuyPriceWithFee = function (collateralAsset, amount, isDesiredInput, payload) {
  let deduceFee = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : true;
  const feeRatio = deduceFee ? _consts.Consts.TBC_FEE : _math.FPNumber.ZERO;
  if (isDesiredInput) {
    const outputAmount = tbcBuyPrice(collateralAsset, amount, isDesiredInput, payload);
    const fee = feeRatio.mul(outputAmount);
    const output = outputAmount.sub(fee);
    const rewards = checkRewards(collateralAsset, output, payload);
    return {
      amount: output,
      fee,
      rewards,
      distribution: [{
        input: collateralAsset,
        output: _consts.Consts.XOR,
        market: _consts.LiquiditySourceTypes.MulticollateralBondingCurvePool,
        income: amount,
        outcome: output,
        fee
      }]
    };
  } else {
    const amountWithFee = (0, _utils.safeDivide)(amount, _math.FPNumber.ONE.sub(feeRatio));
    const inputAmount = tbcBuyPrice(collateralAsset, amountWithFee, isDesiredInput, payload);
    const fee = amountWithFee.sub(amount);
    const rewards = checkRewards(collateralAsset, amount, payload);
    return {
      amount: inputAmount,
      fee,
      rewards,
      distribution: [{
        input: collateralAsset,
        output: _consts.Consts.XOR,
        market: _consts.LiquiditySourceTypes.MulticollateralBondingCurvePool,
        income: inputAmount,
        outcome: amount,
        fee
      }]
    };
  }
};
const tbcBuyPriceNoVolume = (collateralAsset, payload) => {
  const basePriceWrtRef = tbcBuyFunction(collateralAsset, _consts.PriceVariant.Buy, _math.FPNumber.ZERO, payload);
  const collateralPricePerReferenceUnit = tbcReferencePrice(collateralAsset, _consts.PriceVariant.Sell, payload);
  return (0, _utils.safeDivide)(basePriceWrtRef, collateralPricePerReferenceUnit);
};
exports.tbcBuyPriceNoVolume = tbcBuyPriceNoVolume;
const tbcSellPriceNoVolume = (collateralAsset, payload) => {
  const basePriceWrtRef = tbcSellFunction(collateralAsset, _math.FPNumber.ZERO, payload);
  const collateralPricePerReferenceUnit = tbcReferencePrice(collateralAsset, _consts.PriceVariant.Buy, payload);
  return (0, _utils.safeDivide)(basePriceWrtRef, collateralPricePerReferenceUnit);
};
exports.tbcSellPriceNoVolume = tbcSellPriceNoVolume;
const tbcQuoteWithoutImpact = function (inputAsset, outputAsset, amount, isDesiredinput, payload) {
  let deduceFee = arguments.length > 5 && arguments[5] !== undefined ? arguments[5] : true;
  try {
    if ((0, _utils.isXorAsset)(inputAsset)) {
      const xorPrice = tbcSellPriceNoVolume(outputAsset, payload);
      const newFee = deduceFee ? _consts.Consts.TBC_FEE.add(sellPenalty(outputAsset, payload)) : _math.FPNumber.ZERO;
      if (isDesiredinput) {
        const feeAmount = newFee.mul(amount);
        const collateralOut = amount.sub(feeAmount).mul(xorPrice);
        return collateralOut;
      } else {
        const xorIn = (0, _utils.safeDivide)(amount, xorPrice);
        const inputAmountWithFee = (0, _utils.safeDivide)(xorIn, _math.FPNumber.ONE.sub(newFee));
        return inputAmountWithFee;
      }
    } else {
      const xorPrice = tbcBuyPriceNoVolume(inputAsset, payload);
      const feeRatio = deduceFee ? _consts.Consts.TBC_FEE : _math.FPNumber.ZERO;
      if (isDesiredinput) {
        const xorOut = (0, _utils.safeDivide)(amount, xorPrice);
        const feeAmount = xorOut.mul(feeRatio);
        return xorOut.sub(feeAmount);
      } else {
        const outputAmountWithFee = (0, _utils.safeDivide)(amount, _math.FPNumber.ONE.sub(feeRatio));
        const collateralIn = outputAmountWithFee.mul(xorPrice);
        return collateralIn;
      }
    }
  } catch (error) {
    return _math.FPNumber.ZERO;
  }
};
exports.tbcQuoteWithoutImpact = tbcQuoteWithoutImpact;
const tbcQuote = function (inputAsset, outputAsset, amount, isDesiredInput, payload) {
  let deduceFee = arguments.length > 5 && arguments[5] !== undefined ? arguments[5] : true;
  try {
    if ((0, _utils.isXorAsset)(inputAsset)) {
      return tbcSellPriceWithFee(outputAsset, amount, isDesiredInput, payload, deduceFee);
    } else {
      return tbcBuyPriceWithFee(inputAsset, amount, isDesiredInput, payload, deduceFee);
    }
  } catch (error) {
    return (0, _utils.safeQuoteResult)(inputAsset, outputAsset, amount, _consts.LiquiditySourceTypes.MulticollateralBondingCurvePool);
  }
};
exports.tbcQuote = tbcQuote;
