// @sora-substrate/util/build/rewards/consts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RewardingEvents = exports.RewardType = void 0;
let RewardingEvents;
exports.RewardingEvents = RewardingEvents;
(function (RewardingEvents) {
  RewardingEvents["LiquidityProvision"] = "LiquidityProvision";
  RewardingEvents["XorErc20"] = "XorErc20";
  RewardingEvents["SoraFarmHarvest"] = "SoraFarmHarvest";
  RewardingEvents["NftAirdrop"] = "NftAirdrop";
  RewardingEvents["BuyOnBondingCurve"] = "BuyOnBondingCurve";
  RewardingEvents["LiquidityProvisionFarming"] = "LiquidityProvisionFarming";
  RewardingEvents["MarketMakerVolume"] = "MarketMakerVolume";
  RewardingEvents["Unspecified"] = "Unspecified";
})(RewardingEvents || (exports.RewardingEvents = RewardingEvents = {}));
let RewardType;
exports.RewardType = RewardType;
(function (RewardType) {
  RewardType["Provision"] = "Provision";
  RewardType["Strategic"] = "Strategic";
  RewardType["Crowdloan"] = "Crowdloan";
  RewardType["External"] = "External";
})(RewardType || (exports.RewardType = RewardType = {}));
