// @sora-substrate/util/build/bridgeProxy/eth/consts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EthRequestType = exports.EthNetwork = exports.EthCurrencyType = exports.EthAssetKind = void 0;
let EthNetwork;
exports.EthNetwork = EthNetwork;
(function (EthNetwork) {
  EthNetwork[EthNetwork["Ethereum"] = 0] = "Ethereum";
})(EthNetwork || (exports.EthNetwork = EthNetwork = {}));
let EthAssetKind;
/**
 * Type of request which we will wait
 */
exports.EthAssetKind = EthAssetKind;
(function (EthAssetKind) {
  EthAssetKind["Sidechain"] = "Sidechain";
  EthAssetKind["SidechainOwned"] = "SidechainOwned";
  EthAssetKind["Thischain"] = "Thischain";
})(EthAssetKind || (exports.EthAssetKind = EthAssetKind = {}));
let EthRequestType;
/**
 * Type of currency for approval request next steps
 */
exports.EthRequestType = EthRequestType;
(function (EthRequestType) {
  EthRequestType["Transfer"] = "Transfer";
  EthRequestType["TransferXOR"] = "TransferXOR";
  EthRequestType["AddAsset"] = "AddAsset";
  EthRequestType["AddPeer"] = "AddPeer";
  EthRequestType["RemovePeer"] = "RemovePeer";
  EthRequestType["ClaimPswap"] = "ClaimPswap";
  EthRequestType["CancelOutgoingRequest"] = "CancelOutgoingRequest";
  EthRequestType["MarkAsDone"] = "MarkAsDone";
})(EthRequestType || (exports.EthRequestType = EthRequestType = {}));
let EthCurrencyType;
exports.EthCurrencyType = EthCurrencyType;
(function (EthCurrencyType) {
  EthCurrencyType["AssetId"] = "AssetId";
  EthCurrencyType["TokenAddress"] = "TokenAddress";
})(EthCurrencyType || (exports.EthCurrencyType = EthCurrencyType = {}));
